### ISSUE TITLE
[Bug]: MoE topk_softmax/topk_sigmoid return wrong routing weights at (num_experts=128, topk=8)

### ISSUE THREAD

**Issue by sxvvv** [MetaX-MACA/vLLM-metax#270]
### Summary

The MoE gating kernels `_moe_C.topk_softmax` / `_moe_C.topk_sigmoid` return **wrong routing weights at exactly `(num_experts=128, topk=8)`**. Expert *selection* (the returned ids) is correct in all cases — only the **weights** are wrong, and only at this one `(experts, topk)` tile. It looks like a dispatch/template bug that selects the wrong weight branch for this specific shape.

These ops back the generic (non-grouped) fused-MoE routing path inherited from vLLM (`fused_moe.fused_topk` → `ops.topk_softmax`), so `(128, top-8)` — a common MoE shape — silently gets wrong routing weights.

### Observed vs expected (renormalize=False, no bias)

| op | shape | kernel weight returns | expected |
|---|---|---|---|
| `topk_softmax` | (128, 8) | biased weights when a correction bias is passed | unbiased `softmax(g)` |
| `topk_sigmoid` | (128, 8) | `softmax(g)` (wrong activation) | `sigmoid(g)` |

`topk_softmax` with **no** bias is correct even at (128, 8). All other `(experts, topk)` combinations are correct for both ops (verified `experts ∈ {8,16,32,48,64,96,128,192,256}`, `topk ∈ {1..16}`). The defect is **data-independent** (reproduces across seeds) and **token-count-independent**.

### Minimal reproduction

```python
import torch
torch.ops.load_library(
    "/opt/conda/lib/python3.10/site-packages/mcoplib/_moe_C.abi3.so"
)

def gate(opname, gating, topk, bias=None):
    nt = gating.shape[0]
    w   = torch.empty((nt, topk), dtype=torch.float32, device="cuda")
    idx = torch.empty((nt, topk), dtype=torch.int32, device="cuda")
    tei = torch.empty((nt, topk), dtype=torch.int32, device="cuda")
    getattr(torch.ops._moe_C, opname)(w, idx, tei, gating, False, bias)
    return w, idx

torch.manual_seed(0)
for ne, tk in [(64, 8), (128, 8), (256, 8)]:   # only ne=128 is wrong
    g = torch.randn(2, ne, dtype=torch.float32, device="cuda")
    w, idx = gate("topk_sigmoid", g.contiguous(), tk)
    ids = idx[0].long()
    e_sig = (w[0] - torch.sigmoid(g)[0, ids]).abs().max().item()
    e_sm  = (w[0] - torch.softmax(g, -1)[0, ids]).abs().max().item()
    verdict = "OK" if e_sig < 1e-3 else "BUG -> matches softmax path"
    print(f"topk_sigmoid ne={ne:3d} tk={tk}: "
          f"|kernel-sigmoid|={e_sig:.5f}  |kernel-softmax|={e_sm:.5f}  [{verdict}]")
```

Output on C500:

```
topk_sigmoid ne= 64 tk=8: |kernel-sigmoid|=0.00000  |kernel-softmax|=0.80708  [OK]
topk_sigmoid ne=128 tk=8: |kernel-sigmoid|=0.87397  |kernel-softmax|=0.00000  [BUG -> matches softmax path]
topk_sigmoid ne=256 tk=8: |kernel-sigmoid|=0.00000  |kernel-softmax|=0.91188  [OK]
```

### Environment

- MetaX C500, MACA 3.5.3.20, torch 2.8.0+metax3.5.3.9
- `mcoplib` 0.4.2, `_moe_C` GIT_COMMIT e482051, Vllm Op Version 0.17.0

### Test coverage

PR #269 adds `tests/kernels/moe/test_topk_softmax.py` covering these ops. The affected `(128, 8)` cases are marked `xfail(strict=True)` there, so once this kernel is fixed those cases will flip to unexpected-pass and flag the fix — the test does not need editing to pick up the repair, just removal of the `xfail` marker.

Related: roadmap #233, test-porting tracker #254.

**Comment by sxvvv** [comment]
Investigation update (2026-06-07):

I checked the source path for this operator on current `master` and the suspicious shape maps to the optimized `topkGating` dispatch in `csrc/moe/topk_softmax_kernels.cu`. The generic fallback path already exists in the same file:

- `moeSoftmax` / `moeSigmoid`
- followed by `moeTopK`

A narrow source-side mitigation would be to send only `num_experts == 128 && topk == 8` through that existing generic path, and allocate `num_tokens * num_experts` float workspace for that shape. This keeps all other power-of-two expert counts on the current optimized dispatch, so the behavioral blast radius is limited to the bad tile reported here.

I have not opened a fix PR yet because the current rental image runs the precompiled `mcoplib` `_moe_C` extension by default, and source-build validation is not complete. The runtime package used for the repro prints:

- `mcoplib` 0.4.2
- MACA 3.5.3.20
- Vllm Op Version 0.17.0

Source-build blocker on this image:

- `MACA_PATH=/opt/maca` is needed.
- `cmake_maca` exists under `/opt/maca/tools/cu-bridge/tools/cmake_maca`.
- `cmake_maca` then fails because plain `cmake` is missing from PATH / the image.

So the current status is:

- bug reproduced on real C500 runtime
- candidate source fallback identified
- local diff sanity check passes (`git diff --check`)
- source-build and after-fix numerical validation still pending

Once a complete MACA source-build environment is available, the next validation should be:

1. build `_moe_C` from source with the narrow fallback;
2. rerun the minimal repro in this issue;
3. rerun PR #269 and remove the strict `xfail` markers for the fixed `(128, 8)` cases if they turn green.

This is intentionally not a "fix is done" claim yet; the main unresolved part is proving that the source-built extension changes the same runtime path that currently fails.

**Comment by Zymonody7** [comment]
Opened a fix PR for this: #284.

The kernel itself is untouched — since expert ids are correct at every shape, the patch recomputes the weights from `gating_output` at the kernel-selected ids (via the existing `patch/bugfix` mechanism), only at the (128, 8) tile. On first (128, 8) call each op probes the kernel against a reference once and deactivates itself on fixed kernels, so the workaround disappears once mcoplib is fixed.

Verified on C500, mcoplib 0.4.5: E=128 K=8 softmax+bias max_diff 3.03 → 0.0(renormalize both), all other shapes ≤1e-7. Regression tests are included in the PR at the patched ops level; the `_moe_C` kernel-level xfails from #269 stay until the kernel itself is fixed.

**Comment by sxvvv** [comment]
Thanks for picking this up @Zymonody7, and for looping back here to let me know.

The layering makes sense to me — #284 patches `vllm._custom_ops.topk_softmax/sigmoid` (the path real MoE routing goes through), while #269 only exercises the raw `torch.ops._moe_C` op, so the two don't overlap. I'll leave #269 as-is so its strict-xfails stay a sentinel until the underlying kernel is fixed, as you noted. Looking forward to #284 landing.

**Comment by Mandaluoren** [comment]
Seems to already been fixed in the new version.
My test version is:
- mcoplib 0.4.7
- MACA 3.8.0.21
- Vllm 0.22.0
<img width="397" height="440" alt="Image" src="https://github.com/user-attachments/assets/70ff979f-8e46-45c7-801d-499048091676" />
